tests/validation/tests/v3_api/test_vmwarevsphere_driver.py

```diff
@@ -0,0 +1,217 @@
+    ],
+    "type": "vmwarevsphereConfig",
+    "username": "",
+    "vappIpallocationpolicy": "",
+    "vappIpprotocol": "",
+    "vappProperty": "",
+    "vappTransport": "",
+    "vcenter": "",
+    "vcenterPort": "443",
+}
+
+
+def test_valid_environment_variables():
+    assert RANCHER_VSPHERE_USERNAME != '',  \
+        "vSphere User is required to make a cloud credential"
+    assert RANCHER_VSPHERE_PASSWORD != '', \
+        "vSphere Password is required to make a cloud credential"
+    assert RANCHER_VSPHERE_VCENTER != '', \
+        "vCenter URL is required to make a cloud credential"
+
+
+def test_create_cluster():
+    client = get_client_for_token(ADMIN_TOKEN)
+    cloud_cred = create_vsphere_credential(client)
+    nt = create_vsphere_nodetemplate(client, cloud_cred)
+    cluster = client.create_cluster(
+        name=CLUSTER_NAME,
+        rancherKubernetesEngineConfig=rke_config)
+
+    master_pool = client.create_node_pool({
+        "type": "nodetemplate",
+        "clusterId": cluster.id,
+        "controlPlane": True,
+        "etcd": True,
+        "hostnamePrefix": CLUSTER_NAME + "-master",
+        "nodeTemplateId": nt.id,
+        "quantity": 1,
+        "worker": False,
+    })
+
+    worker_pool = client.create_node_pool({
+        "type": "nodetemplate",
+        "clusterId": cluster.id,
+        "controlPlane": False,
+        "etcd": False,
+        "hostnamePrefix": CLUSTER_NAME + "-worker",
+        "nodeTemplateId": nt.id,
+        "quantity": 1,
+        "worker": True,
+    })
+
+    client.wait_success(master_pool)
+    client.wait_success(worker_pool)
+    wait_for_cluster_node_count(client, cluster, 2, 900, 5)
+
+
+def test_nodes_ready():
+    client = get_client_for_token(ADMIN_TOKEN)
+    cluster = get_cluster_by_name(client=client, name=CLUSTER_NAME)
+    validate_cluster_state(client, cluster)
+
+
+def test_ingress():
+    namespace = create_namespace(ADMIN_TOKEN)
+    p_client = namespace["p_client"]
+    ns = namespace["ns"]
+    cluster = namespace["cluster"]
+    con = [{"name": "test1",
+            "image": TEST_IMAGE}]
+    name = random_test_name("default")
+    workload = p_client.create_workload(name=name,
+                                        containers=con,
+                                        namespaceId=ns.id,
+                                        daemonSetConfig={})
+
+    validate_workload(p_client, workload, "daemonSet", ns.name,
+                      len(get_schedulable_nodes(cluster)))
+
+    host = "test1.com"
+    path = "/name.html"
+    rule = {"host": host,
+            "paths": [{"workloadIds": [workload.id], "targetPort": "80"}]}
+    p_client.create_ingress(name=name,
+                            namespaceId=ns.id,
+                            rules=[rule])
+    validate_ingress(namespace["p_client"], namespace["cluster"],
+                     [workload], host, path)
+
+
+def test_cluster_cleanup():
```

Comment:
Put this into a pytest finalizers so it will run even if your tests bomb so clusters don't leak.